onedal/primitives/tree_visitor.cpp

```diff
@@ -153,52 +184,31 @@ to_sklearn_tree_object_visitor<Task>::to_sklearn_tree_object_visitor(std::size_t
     this->max_depth = _depth;
     this->leaf_count = _n_leafs;
     this->class_count = _max_n_classes;
-
-    auto node_ar_shape = py::array::ShapeContainer({ this->node_count });
-    auto node_ar_strides = py::array::StridesContainer({ sizeof(skl_tree_node) });
-
-    auto value_ar_shape = py::array::ShapeContainer({ static_cast<Py_ssize_t>(this->node_count),
-                                                      1,
-                                                      static_cast<Py_ssize_t>(this->class_count) });
-    auto value_ar_strides = py::array::StridesContainer(
-        { this->class_count * sizeof(double), this->class_count * sizeof(double), sizeof(double) });
-
-    skl_tree_node* node_ar_ptr = new skl_tree_node[this->node_count];
-
-
     OVERFLOW_CHECK_BY_MULTIPLICATION(std::size_t, this->node_count, this->class_count);
-    double* value_ar_ptr =
-        new double[this->node_count * 1 *
-                   this->class_count](); // oneDAL only supports scalar responses for now
-
-    this->node_ar = py::array_t<skl_tree_node>(node_ar_shape, node_ar_strides, node_ar_ptr);
-    this->value_ar = py::array_t<double>(value_ar_shape, value_ar_strides, value_ar_ptr);
+    this->node_ar = new skl_tree_node[this->node_count];
```

Comment:
Who owns this memory / how is it cleaned? I don't see any corresponding frees.

Reply:
Maybe also a bug in the daal4py side? I did it to match this: https://github.com/intel/scikit-learn-intelex/blob/master/src/tree_visitor.h#L211

Reply:
In both it is owned on python side.

Reply:
And how/where does the memory get freed and cleaned up?

Reply:
* For `onedal` implementation:
https://github.com/intel/scikit-learn-intelex/blob/master/onedal/primitives/tree_visitor.cpp#L301:L302
See `py::return_value_policy::take_ownership`. Please read about it https://pybind11-jagerman.readthedocs.io/en/stable/advanced.html#return-value-policies

* For `daal4py` implementation:
During daal4py building stage `getree.pyx` copies into `daal4py_cy.pyx` and then cythonizing gives `daal4py_cy.cpp`, find it in build dir.
It is really hard debug cython auto generated c code, I find this code from there.
```c
static void __pyx_tp_dealloc_7daal4py_8_daal4py_pyTreeState(PyObject *o) {
  struct __pyx_obj_7daal4py_8_daal4py_pyTreeState *p = (struct __pyx_obj_7daal4py_8_daal4py_pyTreeState *)o;
  #if CYTHON_USE_TP_FINALIZE
  if (unlikely(PyType_HasFeature(Py_TYPE(o), Py_TPFLAGS_HAVE_FINALIZE) && Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->node_ar);
  Py_CLEAR(p->value_ar);
  (*Py_TYPE(o)->tp_free)(o);
}
```
So it seems that Cython also allows python ownership this.
 
**In any way I am agree that we should check mem leak.** (With valgrind or some other dynamic analysis tools)

Reply:
Okay, especially the `daal4py` approach seems a bit magical. I agree that we should check for mem leaks. @icfaust is this something you could do or do you need help?

Reply:
Just as info @samir-nasibli @ahuber21 , @ahuber21 had it correct in terms of memory leaking.  Unfortunately the pybind11 documentation is extremely limited on the subject. If we look at the pybind11/numpy.h , the way that was implemented using new calls would create a memory leak due to the inherent copy of the array class in pybind11.  https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L747  The way to do it is to pass a dummy base python object, of which py::none() is sufficient. This is definitely a work around that others have faced https://github.com/pybind/pybind11/issues/323

Reply:
Also PyArray_NewFromDescr will not initialize the memory to the default values of the class, so there was a bug in skl_tree_nodes for split nodes where certain values (like threshold) were not set to their defaults.